File: suite/common/permissions/treeUtils.js
/* -*- Mode: Java; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*-
 *
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS
 * IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * rights and limitations under the License.
 *
 * The Original Code is mozilla.org code.
 *
 * The Initial Developer of the Original Code is Netscape
 * Communications Corporation.  Portions created by Netscape are
 * Copyright (C) 1998 Netscape Communications Corporation. All
 * Rights Reserved.
 *
 * Contributor(s):
 */

var setID=0, setValue=0, getID=0, getValue=0;
function Wallet_ColumnSort(columnPosition, childrenName)
{
  //var startTime = new Date();

  // determine if sort is to be ascending or descending
  //    it is ascending if it is the first sort for this tree
  //    it is ascending if last sort was on a different column
  //    else it is the opposite of whatever the last sort for this column was
  var children = document.getElementById(childrenName);
  var lastColumnPosition = children.getAttribute('lastColumnPosition');
  var ascending = true;
  if (columnPosition == lastColumnPosition) {
    ascending = (children.getAttribute('lastAscending') != 'true');
  } else {
    children.setAttribute('lastColumnPosition', columnPosition);
  }
  children.setAttribute('lastAscending', ascending);
  bubbleSort(columnPosition, ascending, children);
  //dump("bubble sort time="+((new Date())-startTime)+"\n");
  return true;
}

// XXX we would like to use Array.prototype.sort, but the DOM doesn't let it
// XXX swap elements using property sets
function bubbleSort(columnPosition, ascending, children)
{
  var a = children.childNodes;
  var n = a.length, m = n - 1;

  // for efficiencey, read all the value attributes only once and store in an array
  var keys = [];
  for (var x=0; x<n; x++){
    var keyCell = a[x].firstChild.childNodes[columnPosition];
    keys[x] = keyCell.getAttribute('value');
  }

  for (var i = 0; i < m; i++) {
    var key = keys[i];
    var winner = -1;
    for (var j = i + 1; j < n; j++) {
      var nextKey = keys[j];
      if (ascending ? key > nextKey : key < nextKey) {
        key = nextKey;
        winner = j;
      }
    }
    if (winner != -1){

      // get the corresponding menuitems
      var item = a[i];
      var row = item.firstChild;
      var nextItem = a[winner];
      var nextRow = nextItem.firstChild;

      // swap the row values and the id's of the corresponding menuitems
      var temp = item.getAttribute('id');
      item.setAttribute('id', nextItem.getAttribute('id'));
      nextItem.setAttribute('id', temp);

      var cell = row.firstChild;
      var nextCell = nextRow.firstChild;
      var position = 0;
      while (cell) {
        var value_i;
        var value_winner;
        if (position == columnPosition) {
          value_i = keys[i];
          value_winner = keys[winner];
          keys[i] = value_winner;
          keys[winner] = value_i;
        } else {
          value_i = cell.getAttribute('value');
          value_winner = nextCell.getAttribute('value');
        }
        cell.setAttribute('value', value_winner);
        nextCell.setAttribute('value', value_i);
        cell = cell.nextSibling;
        nextCell = nextCell.nextSibling;
        position++;
      }
    }
  }
}

File: js/jsdj/classes/com/netscape/jsdebugging/apitests/analyzing/tree/TreeUtils.java
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 2 -*-
 *
 * The contents of this file are subject to the Netscape Public License
 * Version 1.0 (the "NPL"); you may not use this file except in
 * compliance with the NPL.  You may obtain a copy of the NPL at
 * http://www.mozilla.org/NPL/
 *
 * Software distributed under the NPL is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the NPL
 * for the specific language governing rights and limitations under the
 * NPL.
 *
 * The Initial Developer of this code under the NPL is Netscape
 * Communications Corporation.  Portions created by Netscape are
 * Copyright (C) 1998 Netscape Communications Corporation.  All Rights
 * Reserved.
 */


package com.netscape.jsdebugging.apitests.analyzing.tree;

import java.util.*;
import com.netscape.jsdebugging.apitests.analyzing.data.*;

/**
 * Various  utilities for working with trees.
 *
 * @author Alex Rakhlin
 */

public class TreeUtils {
    
    /**
     * Goes through the tree, sets pointers of nodes to point to the objects corresponding
     * to serial numbers, removes the serial number tag altoghether.
     */
    public static void replaceSerialNumbersByPointers (DataPoolManager dpm, TreeNode head, String sn_tag){
        if (head == null) return;
        Vector children = head.getChildren();
        for (int i = 0; i<children.size(); i++){
            TreeNode t = (TreeNode) children.elementAt (i);
            if (t.getTag().equals (sn_tag)) {
                head.setPointer (dpm.lookup (head.getTag(), Integer.valueOf(t.getText()).intValue()));
                t.getParent().removeChild (t);
            }
            else  replaceSerialNumbersByPointers (dpm, t, sn_tag);
        }
    }
    
    /**
     * OK. So, we can have two nodes in the tree point to two different objects in the pools
     * which actually have the same data. When we remove duplicates, we leave only one, and set
     * the "real_pointer" pointer of the second object to point to the first. Now we traverse the
     * tree and change pointers of the nodes to point to the objects which were not removed.
     * After this, duplicate objects, which were removed from pools should be gc'd.
     */
    public static void fixRealPointers (TreeNode head){
        if (head == null) return;
        Vector children = head.getChildren();
        for (int i = 0; i<children.size(); i++){
            TreeNode t = (TreeNode) children.elementAt (i);
            if (t.getPointer () != null) {
                DataSerializable redirect = t.getPointer().getRealPointer();
                if (redirect != null) {
                    //System.out.println ("Redirecting pointer from "+t.getPointer()+" to "+redirect);
                    t.setPointer(redirect);
                }
            }
            fixRealPointers (t);
        }
    }
    
    /**
     * Get the String data of the node
     */
    public static String getStringData (TreeNode head){
        return head.getText();
    }

    /**
     * Get the boolean data of the node
     */
    public static boolean getBooleanData (TreeNode head){
        String s = head.getText();
        if (s.equalsIgnoreCase ("true")) return true;
        return false;
    }

    /**
     * Get the integer data of the node
     */
    public static Integer getIntegerData (TreeNode head){
        Integer i = null;
        String s = head.getText();
        try{
            i = new Integer (s);
        } catch (NumberFormatException e) { System.out.println ("Invalid number format"); }
        return i;
    }
    
    /**
     * Same as above, but returns data of the first child with childTag tag
     */
    public static String getChildStringData (TreeNode head, String childTag){
        TreeNode child = getFirstTagImmediate (head, childTag);
        if (child == null) return null;
        return child.getText();
    }

    /**
     * Same as above, but returns data of the first child with childTag tag
     */
    public static Integer getChildIntegerData (TreeNode head, String childTag){
        TreeNode child = getFirstTagImmediate (head, childTag);
        if (child == null) return null;
        return getIntegerData (child);
    }

    /**
     * Same as above, but returns data of the first child with childTag tag
     */
    public static boolean getChildBooleanData (TreeNode head, String childTag){
        TreeNode child = getFirstTagImmediate (head, childTag);
        if (child == null) return false;
        return getBooleanData (child);
    }

    /**
     * Returns a vector of immediate children with the given tag
     */
    public static Vector searchImmediateChildren (TreeNode head, String tag){
        Vector result = new Vector();
        Vector children = head.getChildren();
        int size = children.size();
        for (int i = 0; i<size; i++){
            TreeNode t = (TreeNode) children.elementAt (i);
            if (t.getTag ().equals (tag)) result.addElement (t);
        }
        return result;
    }
    
    /**
     * Get the first immediate child with the given tag
     */
    public static TreeNode getFirstTagImmediate (TreeNode head, String tag){
        Vector children = head.getChildren();
        int size = children.size();
        for (int i = 0; i<size; i++){
            TreeNode t = (TreeNode) children.elementAt (i);
            if (t.getTag ().equals (tag)) return t;
        }
        return null;
    }
    
    /* Finds first occurence
    */
    public static TreeNode searchDepthFirst (TreeNode head, String tag){
        if (head == null) return null;
        Vector children = head.getChildren();
        for (int i = 0; i<children.size(); i++){
            TreeNode t = (TreeNode) children.elementAt (i);
            if (t.getTag().equals (tag)) return t;
            TreeNode result = searchDepthFirst (t, tag);
            if (result != null) return result;
        }
        return null;
    }
    
    /**
     * Returns all nodes with the specified tag. Children come before parents 
     */
    public static Vector findAllTags (TreeNode head, String tag){
        Vector list = new Vector();
        _findAllTags (head, tag, list);
        return list;
    }
    
    private static void _findAllTags (TreeNode head, String tag, Vector list){
        if (head == null) return;
        Vector children = head.getChildren ();
        for (int i = 0; i<children.size(); i++){
            TreeNode t = (TreeNode) children.elementAt (i);
            _findAllTags (t, tag, list);
            if (t.getTag().equals (tag)) list.addElement (t);
        }
    }
    
    /** 
     * Returns all nodes with the specified tag and specified child tag
     * Leaving tag="" outputs all tags with specified child tag 
     */
    public static Vector findAllTagsWithChild (TreeNode head, String tag, String childTag){
        Vector list = new Vector();
        _findAllTagsWithChild (head, tag, childTag, list);
        return list;
    }
    
    private static void _findAllTagsWithChild (TreeNode head, String tag, String childTag, Vector list){
        if (head == null) return;
        Vector children = head.getChildren ();
        for (int i = 0; i<children.size(); i++){
            TreeNode t = (TreeNode) children.elementAt (i);
            _findAllTagsWithChild (t, tag, childTag, list);
            if (t.getTag().equals (tag) || tag.equals("")) {
                Vector v = t.getChildren();
                if (searchImmediateChildren (t, childTag).size() != 0)
                    list.addElement (t);
            }
        }
    }
    
}